common/src/lib.rs

```diff
@@ -0,0 +1,82 @@
+use tokio::signal;
+
+/// Returns a Future that completes once a signal to shutdown the service is caught.
+pub async fn shutdown_signal() {
+    let ctrl_c = async {
+        signal::ctrl_c() // aka. SIGINT on Unix
+            .await
+            .expect("failed to install Ctrl+C (SIGINT) handler");
+    };
+
+    #[cfg(unix)]
+    let sighup = async {
+        signal::unix::signal(signal::unix::SignalKind::hangup())
+            .expect("failed to install signal handler for SIGHUP")
+            .recv()
+            .await;
+    };
+    #[cfg(not(unix))]
+    let sighup = std::future::pending::<()>();
+
+    #[cfg(unix)]
+    let sigquit = async {
+        signal::unix::signal(signal::unix::SignalKind::quit())
+            .expect("failed to install signal handler for SIGQUIT")
+            .recv()
+            .await;
+    };
+    #[cfg(not(unix))]
+    let sigquit = std::future::pending::<()>();
+
+    #[cfg(unix)]
+    let sigpipe = async {
+        signal::unix::signal(signal::unix::SignalKind::hangup())
+            .expect("failed to install signal handler for SIGPIPE")
+            .recv()
+            .await;
+    };
+    #[cfg(not(unix))]
+    let sigpipe = std::future::pending::<()>();
+
+    #[cfg(unix)]
+    let sigalrm = async {
+        signal::unix::signal(signal::unix::SignalKind::alarm())
+            .expect("failed to install signal handler for SIGALRM")
+            .recv()
+            .await;
+    };
+    #[cfg(not(unix))]
+    let sigalrm = std::future::pending::<()>();
+
+    #[cfg(unix)]
+    let sigterm = async {
+        signal::unix::signal(signal::unix::SignalKind::terminate())
+            .expect("failed to install signal handler for SIGTERM")
+            .recv()
+            .await;
+    };
+    #[cfg(not(unix))]
+    let sigterm = std::future::pending::<()>();
+
+    #[cfg(unix)]
+    let sigchld = async {
+        signal::unix::signal(signal::unix::SignalKind::child())
+            .expect("failed to install signal handler for SIGCHLD")
+            .recv()
+            .await;
+    };
+    #[cfg(not(unix))]
+    let sigchld = std::future::pending::<()>();
+
+    // TODO: add more signals that should result in a shutdown?
+
+    tokio::select! {
+        _ = ctrl_c => {},
+        _ = sighup => {},
+        _ = sigquit => {},
+        _ = sigpipe => {},
+        _ = sigalrm => {},
+        _ = sigterm => {},
+        _ = sigchld => {},
+    }
+}
```

Comment:
I'm not sure we need to listen for all these different signals. `SIGTERM` is the only relevant one, as that's what systemd sends

Reply:
Ok, will reduce it to SIGTERM and SIGINT (Ctrl-C) then. But do you think we should only catch SIGTERM? Hm ... I guess I don't see exactly what use case we are trying to support here. In particular: what exactly do we risk by not catching given signals? Could it for example result in some kind of inconsistency/incorrectness? And/or that processes stick around (allocating resources) when they shouldn't? (not trying to arguing against catching signals here, only interested in understanding things :) )

Reply:
I agree - all good points here. I guess I was kind of hoping/expecting to get some feedback of this kind on this change so I got a chance to learn something. "Systems level" programming considerations like these I encounter only rarely, so it's good to get a reminder (and believe it or not, but the concept of "systemd process management" was new to me (I know what  process is and that you can interrupt it with a signal)). And I absolutely agree that we shouldn't throw in a bunch of signals just in case. Will reduce it to just SIGTERM then, and add a note about why.

Reply:
To be clear about what systemd does:

When stopping a process (with i.e. `systemctl stop` or `systemctl restart`) it sends `SIGTERM` immediately, followed by `SIGKILL` after 90 seconds. The intention is to give programs a short window to gracefully shutdown, then forcefully terminate if they seem to be hanging.

Reply:
Since we're on the subject: Do you have an idea of what could have happened if we didn't catch SIGTERM and passed the fact to with_graceful_shutdown()? Some kind of resource leakage?

Reply:
Data corruption is the concern. Since we do our SQL queries in transactions, the worst data corruption that should be mitigated anyway, but this furthur helps us towards the ideal case where a request from obsinn is either processed in it's entirety and receives a success response, or doesn't change the DB at all.

Reply:
> considerations like these I encounter only rarely

Actually I think this is relevant to other services you're responsible for, like Frost on Kubernetes. Kubernetes handles termination the same way systemd does with `SIGTERM` followed by `SIGKILL`, albeit with a shorter default grace period of 30 seconds:

https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination

Reply:
Yup. In Frost we do implement graceful shutdown by handling these: SIGTERM, SIGINT, SIGHUP, and SIGQUIT. I realize now that the latter two didn't need to be there.